File: scripts/minimax_chat.py
"""Minimal stdlib client for MiniMax M3 text calls.

Uses MiniMax's Anthropic-compatible Messages endpoint (same wire shape as
scripts/enrich_filing.py), so thinking blocks are filtered out and only the
reply text comes back. No third-party dependency, so it runs in any CI job.

Env:
    MINIMAX_API_KEY    required
    MINIMAX_BASE_URL   optional, default https://api.minimax.io/anthropic
    MINIMAX_MODEL      optional, default MiniMax-M3
"""

import json
import os
import time
import urllib.error
import urllib.request

BASE_URL = os.environ.get("MINIMAX_BASE_URL", "https://api.minimax.io/anthropic").rstrip("/")
MODEL = os.environ.get("MINIMAX_MODEL", "MiniMax-M3")


def available():
    return bool(os.environ.get("MINIMAX_API_KEY"))


def chat(system, user, *, max_tokens=4000, temperature=0.2, timeout=120, retries=3):
    """Return the reply text for one system + user turn. Raises on failure."""
    body = json.dumps({
        "model": MODEL,
        "max_tokens": max_tokens,
        "temperature": temperature,
        "system": system,
        "messages": [{"role": "user", "content": user}],
    }, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(f"{BASE_URL}/v1/messages", data=body, headers={
        "x-api-key": os.environ["MINIMAX_API_KEY"],
        "anthropic-version": "2023-06-01",
        "Content-Type": "application/json",
    }, method="POST")
    for attempt in range(retries + 1):
        try:
            with urllib.request.urlopen(req, timeout=timeout) as r:
                data = json.loads(r.read().decode("utf-8"))
            break
        except urllib.error.HTTPError as e:
            detail = e.read().decode("utf-8", "ignore")[:300]
            if (e.code == 429 or e.code >= 500) and attempt < retries:
                wait = int(e.headers.get("Retry-After") or 2 ** (attempt + 2))
                print(f"MiniMax HTTP {e.code}, retrying in {wait}s...")
                time.sleep(wait)
                continue
            raise RuntimeError(f"MiniMax API {e.code}: {detail}") from e
    text = "".join(b.get("text", "") for b in data.get("content", [])
                   if b.get("type") == "text").strip()
    if not text:
        raise RuntimeError(f"MiniMax returned no text (stop_reason={data.get('stop_reason')})")
    return text

File: scripts/minimax_chat.mjs
// Minimal MiniMax M3 client for the Node eval/feedback scripts. Same wire shape
// as scripts/minimax_chat.py (Anthropic-compatible Messages endpoint), so only
// text blocks come back and thinking blocks are dropped.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const HERE = dirname(fileURLToPath(import.meta.url));
const BASE_URL = (process.env.MINIMAX_BASE_URL || "https://api.minimax.io/anthropic").replace(/\/$/, "");
export const MINIMAX_MODEL = process.env.MINIMAX_MODEL || "MiniMax-M3";

// Env first, then the same .env files the scripts already read the chat token from.
export function loadMinimaxKey() {
  if (process.env.MINIMAX_API_KEY) return process.env.MINIMAX_API_KEY.trim();
  for (const p of [join(HERE, "..", "..", "AI Agent", ".env"), join(HERE, "..", ".env")]) {
    try { const m = readFileSync(p, "utf8").match(/^MINIMAX_API_KEY\s*=\s*(.+)$/m); if (m) return m[1].trim(); }
    catch { /* next */ }
  }
  return null;
}

export async function minimaxChat(key, system, user, { maxTokens = 2000, temperature = 0.2 } = {}) {
  const r = await fetch(`${BASE_URL}/v1/messages`, {
    method: "POST",
    headers: { "Content-Type": "application/json", "x-api-key": key, "anthropic-version": "2023-06-01" },
    body: JSON.stringify({
      model: MINIMAX_MODEL, max_tokens: maxTokens, temperature, system,
      messages: [{ role: "user", content: user }],
    }),
  });
  const d = await r.json().catch(() => ({}));
  if (!r.ok) throw new Error(`MiniMax ${r.status}: ${JSON.stringify(d).slice(0, 200)}`);
  const text = (d.content || []).filter((b) => b.type === "text").map((b) => b.text).join("").trim();
  if (!text) throw new Error(`MiniMax returned no text (stop_reason=${d.stop_reason})`);
  return text;
}

// Model replies sometimes wrap JSON in markdown fences.
export function parseJsonReply(text) {
  let t = text.trim();
  if (t.startsWith("```")) t = t.split("```")[1].replace(/^json/, "").trim();
  return JSON.parse(t);
}
